helios-client/src/main/java/com/spotify/helios/client/AuthenticatingHttpConnector.java

```diff
@@ -0,0 +1,178 @@
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLSession;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+
+/**
+ * HttpConnector that wraps another connector to add the necessary hooks into HttpURLConnection for
+ * our SSH/TLS-based authentication.
+ */
+public class AuthenticatingHttpConnector implements HttpConnector {
+
+  private static final Logger log = LoggerFactory.getLogger(AuthenticatingHttpConnector.class);
+
+  // TODO (mbrown): username is here to pass to SshAgentSSLSocketFactory, could be handled nicer
+  private final String user;
+  private final Optional<AgentProxy> agentProxy;
+  private final List<Identity> identities;
+  private final EndpointIterator endpointIterator;
+
+  private final HttpConnector delegate;
+
+  public AuthenticatingHttpConnector(final String user,
+                                     final Optional<AgentProxy> agentProxy,
+                                     final List<Identity> identities,
+                                     final EndpointIterator endpointIterator,
+                                     final HttpConnector delegate) {
+    this.user = user;
+    this.agentProxy = agentProxy;
+    this.identities = identities;
+    this.endpointIterator = endpointIterator;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public HttpURLConnection connect(final URI uri, final String method, final byte[] entity,
+                                   final Map<String, List<String>> headers) throws HeliosException {
+    final Endpoint endpoint = endpointIterator.next();
+    final String endpointHost = endpoint.getUri().getHost();
+
+    // convert the URI whose hostname portion is a domain name into a URI where the host is an IP
+    // as we expect there to be several different IP addresses besides a common domain name
+    final URI ipUri;
+    try {
+      ipUri = toIpUri(endpoint, uri);
+    } catch (URISyntaxException e) {
+      throw new HeliosException(e);
+    }
+
+    final Deque<Identity> ids;
+    if (ipUri.getScheme().equalsIgnoreCase("https")) {
+      ids = Queues.newArrayDeque(identities);
+    } else {
+      ids = Queues.newArrayDeque();
+    }
+
+    try {
+      while (true) {
+        final Identity identity = ids.poll();
+
+        try {
+          log.debug("connecting to {}", ipUri);
+
+          final HttpURLConnection connection = delegate.connect(ipUri, method, entity, headers);
+
+          handleHttps(connection, endpointHost, identity);
+
+          final int responseCode = connection.getResponseCode();
+          if (((responseCode == HTTP_FORBIDDEN) || (responseCode == HTTP_UNAUTHORIZED))
+              && !ids.isEmpty()) {
```

Comment:
bug found as a result of now having a test here: this was `!identities.isEmpty()` in b02b0a5, but since a copy of the list held as a field is made in `Deque<Identity>` the original collection is always non empty. A NPE would be thrown on the next log line when accessing `identity.getComment()` as `identity` was null when number of identities < number of endpoints.